support-models/src/main/scala/com/gu/support/workers/GiftRecipient.scala

```diff
@@ -54,6 +55,52 @@ object GiftRecipient {
     ))
 }
 
+case class GiftCode private(value: String) extends AnyVal
+
+object GiftCode {
+
+  def apply(value: String): Option[GiftCode] =
+    Some(value)
+      .filter(_.matches(raw"""gd(03|06|12)-[a-km-z02-9]{8}"""))
+      .map(new GiftCode(_))
+
+  implicit val e1 = Encoder.encodeString.contramap[GiftCode](_.value)
+  implicit val d1 = Decoder.decodeString.emap(GiftCode(_).toRight("invalid gift code"))
+
+}
```

Comment:
GiftCode object moved to models, so we can serialise it between steps

Reply:
I wonder whether we actually need a separate type for `GiftCode` we could just use `RedemptionCode` for everything.

Reply:
good point, I did vaguely wonder that at one point.  I will look and decide if to put that change in this PR.

Reply:
I had a look, I think there is a case for sharing the data types in terms of reducing the number of classes, on the other side there is no real logical benefit of them being compatible, and some potential confusion as they are treated totally separately in the functions.
redemption code is for validating the untrusted input from the user, whereas the gift code is the generated one that will be written into zuora.
I propose rename `GiftCode` to `GeneratedGiftCode` in this PR, then it's clear the purpose.  

But overall it's not something I feel strongly about keeping separately so happy to change it again or in future.